### ISSUE TITLE
v1.0.2k: memory leak in function X509_PKEY_new

### ISSUE THREAD

**Issue by pigeon8** [openssl/openssl#3349]
Source code:
```
    104:    X509_PKEY *X509_PKEY_new(void)
    105:    {
    106:        X509_PKEY *ret = NULL;
    107:        ASN1_CTX c;
    108:
    109:        M_ASN1_New_Malloc(ret, X509_PKEY);
    110:        ret->version = 0;
    111:        M_ASN1_New(ret->enc_algor, X509_ALGOR_new);
    112:        M_ASN1_New(ret->enc_pkey, M_ASN1_OCTET_STRING_new);
    113:        ret->dec_pkey = NULL;
    114:        ret->key_length = 0;
    115:        ret->key_data = NULL;
    116:        ret->key_free = 0;
    117:        ret->cipher.cipher = NULL;
    118:        memset(ret->cipher.iv, 0, EVP_MAX_IV_LENGTH);
    119:        ret->references = 1;
    120:        return (ret);
    121:        M_ASN1_New_Error(ASN1_F_X509_PKEY_NEW);
    122:    }
```

Preprocessed code:
```
    104:    X509_PKEY *X509_PKEY_new(void)
    105:    {
    106:        X509_PKEY *ret = ((void *)0);
    107:        ASN1_CTX c;
    108:
    109:        if ((ret=(X509_PKEY *)CRYPTO_malloc((int)sizeof(X509_PKEY),"../crypto/asn1/x_pkey.c",109)) == ((void *)0)) { c.line=109; goto err2; };
    110:        ret->version = 0;
    111:        if (((ret->enc_algor)=X509_ALGOR_new()) == ((void *)0)) return(((void *)0));
    112:        if (((ret->enc_pkey)=(ASN1_OCTET_STRING *) ASN1_STRING_type_new(4)) == ((void *)0)) return(((void *)0));
    113:        ret->dec_pkey = ((void *)0);
    114:        ret->key_length = 0;
    115:        ret->key_data = ((void *)0);
    116:        ret->key_free = 0;
    117:        ret->cipher.cipher = ((void *)0);
    118:        memset(ret->cipher.iv, 0, 16);
    119:        ret->references = 1;
    120:        return (ret);
    121:        err2: ERR_put_error(13,((173)),((1|64)),"../crypto/asn1/x_pkey.c",(c.line)); return(((void *)0));
    122:    }
```

The issue:
If call of X509_ALGOR_new() from line 111 is failed then we existing from the function with loosing allocation of 'ret' (line 109).
If call of ASN1_STRING_type_new() from line 112 is failed then we existing from the function with loosing of two allocations - 'ret' (line 109) and 'ret->enc_algor' (line 111).

**Comment by bernd-edlinger** [comment]
Ping @levitte ?

**Comment by FdaSilvaYY** [comment]
I remember it was fixed int the trunk before 1.1.0 release.
It can't either be backport or fixed without throwing the macro bunch .

**Comment by bernd-edlinger** [comment]
Okay, then I will have a look.

**Comment by bernd-edlinger** [comment]
Fixed on 1.0.2 by 95f966b.
Thanks!
